### PR TITLE
Bug 1253130: added admin-cli file

### PR DESCRIPTION
**Pull request by bfallonf**
As per BZ 1253130:

https://bugzilla.redhat.com/show_bug.cgi?id=1253130

Added a file similar to the basic CLI operations file, but for admin commands. Also changed the current file to "Developer CLI Operations".

@tpoitras @adellape @tnguyen-rh @CowboysFan anyone want to peer review?

### CODE REVIEW

**Comment by adellape** [comment]
@bfallonf Awesome, thanks for starting this. We've been needing it for a long time. I'll try to review today.

**Comment by bfallonf** [comment]
@tpoitras @adellape @tnguyen-rh @CowboysFan @ahardin-rh ping

Any thoughts on this yet? Might be ok to merge as is and we can add to it over time.

**Comment by tnguyen-rh** [comment]
@bfallonf
I like the use of tables, but i don't like the multiple redundancies:
- horizontally: *COMMAND* repeated in Syntax column;
- vertically: `oadm` constant in Syntax column.
These redundancies are particularly annoying because whereas the intro does indeed describe the common syntax of the commands, the tables don't take advantage of that commonality.

Perhaps a better presentation can be achieved if we:
- Say "Common Syntax" instead of "Common Operations" in intro.
- Eliminate Operation column.
- Remove `oadm` from Syntax column.
- Rename Syntax column to Operation.

This would also improve Description readability since, generally, reducing the number of columns in a table allows it to be displayed more concisely.  (thinking out loud...)  I suppose a more radical change would be to use a [labeled list](http://asciidoctor.org/docs/user-manual/#labeled-list)...

**Comment by bfallonf** [comment]
@tnguyen-rh 

I preeeety much just copied off the "oc commands for developers" section we already have [1]. I do get what you're saying, though. Now that you've pointed it out, I can see the repetitiveness. 

I'm not opposed to labeled lists. The point of the tables is great readability, and if that's hindering rather than helping it might be good to change it up. Should we take a vote?

[1] https://docs.openshift.com/enterprise/3.1/cli_reference/basic_cli_operations.html

**Comment by tnguyen-rh** [comment]
@bfallonf 
The commands are readable, but i find the Description barely readable.  Anyway, this particular style issue is bigger than just this PR; we can resolve it asynchronously and follow up w/ these particular tables, later, if need be.  Apart from that, LGTM.

**Comment by adellape** [comment]
@bfallonf @tnguyen-rh The use of tables in basic_cli_operations has always bothered me, mainly for the reasons @tnguyen-rh already mentioned, especially as it's grown and the commands have become more complex. For example, look at how unreadable and jumbled up the Syntax columns gets on the Customer Portal build as you continue down the page:

https://access.redhat.com/documentation/en/openshift-enterprise/3.1/cli-reference/chapter-4-cli-operations

I think this would be a good opportunity to workshop a new layout finally, one that looks good in both builds (a big part of the problem is that in-line code using backticks don't wrap). I think labeled lists is a good option to try, and we could backport the layout changes to basic_cli_operations.

**Comment by ahardin-rh** [comment]
@adellape , @bfallonf , @tnguyen-rh , +1 for trying out a new layout to boost readability.

**Comment by bfallonf** [comment]
@tnguyen-rh @adellape @ahardin-rh Ok. Made some changes. Put the examples into labeled lists and put in little example boxes with the examples. 

I like it. Anyone else have thoughts?

**Comment by tpoitras** [comment]
Late to the conversation here.

I'm a big fan of tables, simply for how quickly I can scan across the grid lines and quickly find information. However, it's terrible the way those tables are rendering in the portal docs (without lines), and how they're centering the text of the first column.

I think these labeled lists look clean, are easy to read, and are an improvement. 

However, I see a problem with the equal whitespace on either side of the example command. Maybe it's just the way that it's rendering here on GitHub. I'm going to ask @bfallonf if I can see a pretty rendered build of this page. :penguin:

**Comment by tpoitras** [comment]
OK, the spacing is not an issue in the actual rendered build. All good. LGTM! Good work, @bfallonf! :+1: :hamburger: :neckbeard:

**Comment by tnguyen-rh** [comment]
@bfallonf LGTM.

**Comment by ahardin-rh** [comment]
@bfallonf  Looks good!  :shipit:

File: cli_reference/admin_cli_operations.adoc [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+This topic provides information on the administrator CLI operations and their syntax. You must link:get_started_cli.html[setup and login] with the CLI before you can perform these operations.
```
Comment (adellape) [inline cli_reference/admin_cli_operations.adoc]: Even if it's pretty obvious, we don't actually ever say anywhere plainly "the administrator CLI (is? / uses?) the oadm command" or similar. Seems like the Overview would be a good place to make that clear from the jump. Same is true of the basic_cli topic; it doesn't actually say "the OpenShift CLI (is? / uses?) the oc command", it just talks about "the CLI" then starts mentioning oc.

Again, it's fairly obvious in both topics, but it seems like a better intro would say so more clearly.

File: cli_reference/admin_cli_operations.adoc [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+[[common-operations]]
+
+== Common Operations
+The CLI allows interaction with the various objects that are managed by
```
Comment (adellape) [inline cli_reference/admin_cli_operations.adoc]: I know this was just carried over from the basic_cli topic, but maybe it should be more specific here:

s/CLI/administrator CLI/

File: cli_reference/admin_cli_operations.adoc [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+OpenShift. Many common `oadm` operations are invoked using the following syntax:
+
+----
+$ oadm <action> <option>
+----
+
+This specifies:
+
+- An `_<action>_` to perform, such as `new-project` or `router`.
+- An available `_<option>_` to perform the action on as well as a value for the option. Options include `--output` or `--credentials`.
```
Comment (adellape) [inline cli_reference/admin_cli_operations.adoc]: The underscores above in `_<action>_` and `_<option>_` look like holdovers from back when we were doing that for replaceables. Can be removed now, I think.

File: cli_reference/admin_cli_operations.adoc [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+[[basic-admin-cli-operations]]
+
+== Basic CLI Operations
+
+`new-project`:: 
+Creates a new projects:
```
Comment (adellape) [inline cli_reference/admin_cli_operations.adoc]: s/projects/project/

File: cli_reference/admin_cli_operations.adoc [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+----
+
+[[maintenance-cli-operations]]
+
+== Maintenance CLI Operations
+
+`build-chain`::
+Outputs the inputs and dependencies of any builds:
+----
+$ oadm build-chain [IMAGESTREAM:TAG]
```
Comment (adellape) [inline cli_reference/admin_cli_operations.adoc]: The tag is optional, so, could be:

```
$ oadm build-chain <image-stream>[:<tag>]
```

File: cli_reference/admin_cli_operations.adoc [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+`version`::
+Displays the version of the indicated object:
+----
+$ oadm version
+----
+
+`help`::
+Displays help about any command:
+----
+$ oadm help
```
Comment (adellape) [inline cli_reference/admin_cli_operations.adoc]: Maybe show:

```
$ oadm help <command>
```

**Comment by adellape** [comment]
This looks loads better than our old tables, thanks for doin it, Brice. :+1: 

@bfallonf @tpoitras @tnguyen-rh @ahardin-rh How do we feel about how the markup is rendered for the commands? Since item labels by default get bolded, when the commands have backticks added to them they come out looking like bold + code (and red, at least in the docs.openshift build), which we normally use for variable names. I'm not sure exactly how that would look on the Portal build, and I'm not sure yet how to (if we even can) do only a local or stage Portal build.

Another option that would address multiple birds w/ one stone would be to maybe make the command names be Level 2 Sections (i.e, `===`) instead of labels. Doing so would:

- (For the docs.openshift builds) Expose the commands in the TOC at the top for easier jumping.
- (For the Portal builds) Expose the commands in the left-hand nav and PDF TOC for easier jumping.
- Create a natural anchor/link point for each command for cross-referencing.

Possible cons:

- We'd be just swapping one non-standard rendering (i.e., bold+code) for another (i.e., however Level 3 Section titles are rendered) for these command names.
- The TOC may get quite long, but maybe that's fine. Seems no less unwieldy than scrolling through our current, long tables. And at least it'd be more useful than our current TOC; for example, as it stands now, how do I know what's gonna be in Settings vs Advanced vs Other? (I know those were just carried over from the oadm output groupings, which I think are fine.)

I feel like the pros outweigh the cons here, but WDYT?

**Comment by adellape** [comment]
@bfallonf I think there's also opportunity to link from each command to their respective topics that show actual usage (for example, `manage-node` to the Managing Nodes topic). I could do that in a follow-up, or whichever.

**Comment by ahardin-rh** [comment]
@bfallonf @adellape Even though the TOC risks getting lengthy, I think turning the command names  into level 2 sections would add value. Good idea.

**Comment by bfallonf** [comment]
Ok. I like the level 2 section thingey, so I've done it up so we can look at it. It works. Not sure about the links out to individual topics though. Not sure if it's worth it. Also added a little more context to the intro para about the oc/oadm command.

Let me know what y'all think. I can make the same changes to the developer guide as well.

**Comment by tpoitras** [comment]
I also like the Lvl 2 section layout, however, I think we might get some pushback about using commands as section headings. 

I especially like it because at a glance of the ToC you immediately can see what the "basic", "settings", "advanced", or "other" commands are all about, since those ones are a bit cryptic.

**Comment by tnguyen-rh** [comment]
@adellape I like that the TOC becomes more useful.  However...

Only objection would be some vague (perhaps false) memory of "no code or code fragments in titles" in a style guide somewhere.  I think the rationale was that titles are subject to automatic transforms (e.g., upcasing) that can result in inaccuracy.

For example, a title "ls -lR" could be munged to actually display as "LS -LR" which is a different (and only partially invalid) command.  A reader could probably infer that the command needs to be downcased, but they might still get fooled by mixed-case option (both "-LR" and "-lr" lose the original meaning).  Of course, the actual non-TOC text would have the proper unmolested command, but you can't count on people actually reading that part as the TOC already has all the info they need (or so they think).

Admittedly this is a contrived example, but i think there is a real risk that people who are too lazy or literal might read it the Wrong Way, w/ the result being a spate of silly BZs whose resolution will eventually return us to the current state.

**Comment by adellape** [comment]
> Not sure about the links out to individual topics though. Not sure if it's worth it.

@bfallonf Can you expand on why it doesn't seem worth it? My thinking was that if you are looking up a command and want to know more, there'd be a link right there to relevant further documentation. Versus hunting or using the search bar (both of which are less elegant, and the latter only available on docs.openshift). The commands as they are in this topic currently only sorta useful; they give a very basic syntax and a brief description, but take `oadm router` or `oc policy` for example. There are whole topics dedicated to everything you can do w/ those.

> however, I think we might get some pushback about using commands as section headings.

@tpoitras Yeah, I figured going into the suggestion that there could be pushback for style reasons, which we 99% of the time adhere to, but since this is supposed to be a "reference" topic, it seems like a perfect use case for being an exception. (Speaking generally) I would for hate for our hands to be tied by something like that at the expense of the topic being more useful.

> Admittedly this is a contrived example, but i think there is a real risk that people who are too lazy or literal might read it the Wrong Way, w/ the result being a spate of silly BZs whose resolution will eventually return us to the current state.

@tnguyen-rh While the upcasing concern is valid (Portal only... and I wish they didn't), I think we're safe here since we're only naming partial commands/subcommands anyway in the heading, which would be unhelpful to the lazy, followed almost immediately by copy/paste-able (sorta) syntax. I say sorta because we show replaceables, so the overly-literal would be doomed either way.

**Comment by CowboysFan** [comment]
@adellape @bfallonf @tpoitras @ahardin-rh 
It's looking pretty good now. As for the linking to other topics, for what it's worth our customers are already complaining that we're linking to too many different topics from one topic. In some places we have several links as part of one sentence, and it's causing a lot of confusion for customers. I think our links need to be more carefully planned and we shouldn't link out if it doesn't relate to the current topic. 

I think this is a good start towards solid reference content for the CLI. So, my vote is to not link to the different topics from the CLI reference; I don't think that will help customers. 

Whatever explanation/info we have to provide customers for using a command should be done in the reference doc itself in my opinion.

**Comment by bfallonf** [comment]
@CowboysFan @adellape Yeah I agree that the links might get too much. I think my reasoning was more along the lines that customers probably wouldn't be looking at this topic first, then wanting to hunt down the topics. If anything, it might be worth putting more effort into the little description sentences.

@tnguyen-rh @tpoitras I think having command examples in this context is ok. It'd be hard to give it another title without losing the reason why this topic exists.

**Comment by tpoitras** [comment]
Good stuff. I like it as it is now. LGTM. :+1: :vhs:

**Comment by adellape** [comment]
@CowboysFan @bfallonf I had seen that comment from the feedback doc but interpreted it more about the over-linking linking within sentences/paragraph prose (as it mentions), not necessarily for quick "See x for more" linking of a quick reference blurb to a relevant, dedicated topic. But, whatever.

> Whatever explanation/info we have to provide customers for using a command should be done in the reference doc itself in my opinion.

Can you clarify this more? So, fully detail all options? And how/when/why to use them? Only some options? Where is the line? Take the `oadm router` command as an example as mentioned in my last comment. How much information would we duplicate between this reference and that topic? It makes me wonder how large this topic could grow.

**Comment by CowboysFan** [comment]
@adellape Not to cause any more confusion, I think as long as customers get a good picture of what the command does and the syntax we're good. I agree with @bfallonf that there's no need to link to other topics in a CLI reference.

I think this topic looks good the way it is now.

**Comment by bfallonf** [comment]
@adellape A few of us in Brisbane chatted about this for a bit. I think the dream is to have something similar to the RHEL cheatsheet that came out not too long ago. Obviously links don't work on printed out pages... But more to it, it depends on the flow of reading. If readers aren't getting a good enough understanding of the commands, then this might be a good topic to look at, after they've been to the other topics.

Either way, I'll do up the developer commands to look the same as these, then that might be all for this BZ. Bilhar has mentioned that these topics might fit into the restructure he's been chatting to us about, so they might get some further tooling in the near future.

**Comment by bfallonf** [comment]
K. Updated this for the dev commands as well.

File: cli_reference/basic_cli_operations.adoc [hunk 7]
```diff
@@ -89,324 +88,324 @@ point the `oc` client to the correct API endpoint. For example: `oc get svc --ap
 == Basic CLI Operations
```
Comment (adellape) [inline cli_reference/basic_cli_operations.adoc]: I feel like now the "Object Types" section (which is currently at the very bottom of the topic) should maybe get slide up to here, between "Common Operations" and "Basic CLI Operations". Sets the understanding of the available objects types and their abbrevs, before moving on to the commands that actually operate on them.

File: cli_reference/basic_cli_operations.adoc [hunk 8]
```diff
@@ -89,324 +88,324 @@ point the `oc` client to the correct API endpoint. For example: `oc get svc --ap
...
+$ oc types
+----
 
-|`logout`
-|`oc logout`
-|End the current session.
+=== `login`
+Log in to the OpenShift server:
+----
+$oc login
```
Comment (adellape) [inline cli_reference/basic_cli_operations.adoc]: Space after $

File: cli_reference/basic_cli_operations.adoc [hunk 9]
```diff
@@ -89,324 +88,324 @@ point the `oc` client to the correct API endpoint. For example: `oc get svc --ap
...
+----
+Edit the desired object type with a specified text editor:
+----
+$ OC_EDITOR="<text_editor>" oc edit <object_type>/<object_type_name>
+----
+Edit the desired object in a specified format (eg: JSON):
+----
+$ oc edit <object_type>/<object_type_name> \
+--output-version=<object_type_version> \
+-o <object_type_format>
```
Comment (adellape) [inline cli_reference/basic_cli_operations.adoc]: Add 4 spaces to the beginning of the returned lines (162 and 163).

File: cli_reference/basic_cli_operations.adoc [hunk 10]
```diff
@@ -89,324 +88,324 @@ point the `oc` client to the correct API endpoint. For example: `oc get svc --ap
...
+----
+Re-run build with name `<build_name>`:
+----
+$ oc start-build --from-build=<build_name>
+----
+Archive `<dir_name>` and build with it as the binary input:
+----
+$ oc start-build --from-dir=<dir_name>
+----
+Use <file_name> as the binary input for the build. This file must be the only one in the build source. For example, *_pom.xml_* or *_Dockerfile_*.
```
Comment (adellape) [inline cli_reference/basic_cli_operations.adoc]: Mark up <file_name> in backticks.

File: cli_reference/basic_cli_operations.adoc [hunk 11]
```diff
@@ -89,324 +88,324 @@ point the `oc` client to the correct API endpoint. For example: `oc get svc --ap
...
+$ oc new-build .
+----
 
-|`import-image`
-|`oc import-image <imagestream>`
-|Import tag and image information from an external Docker image repository.
+=== `cancel-build`
+Stop a build that is in progress:
+----
+$ `oc cancel-build <build_name>`
```
Comment (adellape) [inline cli_reference/basic_cli_operations.adoc]: Take off backticks.

File: cli_reference/basic_cli_operations.adoc [hunk 12]
```diff
@@ -89,324 +88,324 @@ point the `oc` client to the correct API endpoint. For example: `oc get svc --ap
...
-[[troubleshooting-and-debugging-cli-operations]]
+=== `run`
+Create and run a particular image, possibly replicated. Create a deployment configuration to manage the created container(s). You can choose to run in the foreground for an interactive container execution.
+----
+$ oc run NAME --image=<image> \
+[--port=<port>] \
+[--replicas=<replicas>] \
+[--dry-run=<bool>] \
+[--overrides=<inline-json>] \
+[options]
```
Comment (adellape) [inline cli_reference/basic_cli_operations.adoc]: Add 4 spaces to the returned lines after the beginning of the cmd (345 - 349).

File: cli_reference/admin_cli_operations.adoc [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+----
+$ oadm <action> <option>
+----
+
+This specifies:
+
+- An `<action>` to perform, such as `new-project` or `router`.
+- An available `<option>` to perform the action on as well as a value for the option. Options include `--output` or `--credentials`.
+
```
Comment (adellape) [inline cli_reference/admin_cli_operations.adoc]: Same comment here now in the Admin CLI topic, about considering moving "Object Types" from the very bottom of the topic to here, between "Common Operations" (which sets up the most basic form of how the commands are structured) and "Basic CLI Operations" (which begins showing commands that make use of the object types).

**Comment by adellape** [comment]
@bfallonf Comments inline, looks good overall. BTW, I made an erroneous comment in the Admin CLI topic just now, and then deleted it. Just FYI in case you get an email alert about it and are confused as to where it went.

File: cli_reference/basic_cli_operations.adoc (hunk 7 above)
Comment (bfallonf) [inline cli_reference/basic_cli_operations.adoc]: I agree. Done.

**Comment by bfallonf** [comment]
@adellape Cool. Edits made to suggestion. If there's nothing else, this can probably be merged.

**Comment by adellape** [comment]
@bfallonf LGTM after a squarsh.

**Comment by bfallonf** [comment]
@adellape Blah, did that, but must've missed one. I'll merge when Travis is ok with it.

**Comment by adellape** [comment]
Added the dedicated-3.1 label to this, then followed up with https://github.com/openshift/openshift-docs/pull/1511 to make it Dedicated-worthy.

**Pull request merged**